### ISSUE TITLE
Read a resource value without Transition or Suspense

### ISSUE THREAD

**Issue by xbb** [leptos-rs/leptos#1837]
**Is your feature request related to a problem? Please describe.**
At the moment a resource should be read inside a `<Transition />` or `<Suspense />` component to avoid hydration issues and you get a warning when not doing that.

However it's possible that a component with a resource is nested within one of those blocks, for example:

```rust
<Suspense>
    <Component>
       <SubComponent />
   </Component>
</Suspense>
```

If you create and read a resource in `SubComponent` you'll get no warning and the whole parent `Suspense` block will refresh when the resource is refreshed. This might be by design, but it could potentially go unnoticed by new users (like me!) who may not immediately identify the issue.

Also, creating derived signals from resources and passing those around, is error prone, as those also should be wrapped in `Transition`/`Suspense` blocks.

**Describe the solution you'd like**
It would be nice to have an easy and safe way to read a `Resource` value or create a derived signal from it without requiring or interfering with `Transition`/`Suspense`.

**Comment by gbj** [comment]
I agree that it would be nice! I hope it's helpful if I explain why it is the way it is currently.

tl;dr: Suspense is good when you're doing SSR, doing what you're looking for is hard

## When you *can* read a resource without Suspense 
First off, note that you *can* read a resource as if it were a `Signal<Option<T>>` currently, with no problems, if you're either
1) using client-side rendering, not server-side rendering and hydration,
2) using SSR and hydration but this is a subsequent navigation, not the initial page load; (this will emit a warning but should work fine) or
3) using `create_local_resource`

In any of those three cases, `Suspense` and `Transition` just manage the "show a loading state" experience.

## Why you *can't* read a resource without Suspense

The similarity that unites the three cases above is that they're all reading the resource and doing the rendering in the browser. When you're talking about just the client side, rendering directly to the DOM, a resource does in fact work like any other signal, and doesn't require Suspense.

However, the reason Suspense is important with SSR/hydration is that it manages the whole process of async rendering/HTML streaming. If you read a resource outside a Suspense during SSR, the server has no way of knowing when any given chunk of the page is ready to render and send to the client. Suspense is there to give you fine-grained control over which chunks the page will be sent over the network in, and to ensure that fallbacks are replaced correctly in the client.

## Possible Solutions

### For the example given

You gave the example
```
<Suspense>
    <Component>
       <SubComponent />
   </Component>
</Suspense>
```
And note that if you read a resource in `SubComponent`, the whole block will fall back to the `Suspense` fallback while it's loading. Correct! This is exactly what's supposed to happen. It won't re-render, per se; i.e., your Component function shouldn't actually run a second time, and if it does that might be a bug on the framework end. But right, the whole thing will fall back.

There are two solutions here:
1) Use a finer-grained Suspense with its own fallback around the area you actually want to disappear, rather than the coarser-grained one around this whole part of the tree
2) Use Transition here, instead of Suspense, so that Component will stay visible the whole time.

### For the problem in general

#### Wait for resources, instead of Suspenses

I can imagine a hypothetical world in which, if you're using `SsrMode::Async`, Suspense is totally optional and the server simply waits for *all* resources to be loaded before it sends the response, rather than waiting for all Suspense fragments to be ready before it sends the response, on the assumption that you won't create resources you don't use. This might actually be a nice improvement and is worth leaving the issue open for!

I can't see how that would work as well with streaming HTML, which is the default mode, because Suspense is there for you to be able to mark off the chunks. But that doesn't mean it can't be done for async rendering.

#### Default to creating a tiny micro-Suspense

I can also imagine a hypothetical in which, if you read a resource without a Suspense, the framework injects a tiny Suspense around... what? Thinking out loud, here. Around just whatever dynamic wrapper you're reading it in, I guess?

The issue here is that it specifically *wouldn't* fix your example, because in your example you *are* reading within a Suspense, you just end up not liking where you put the Suspense.

## Acknowledging the pain of resource-derived signals

Just want to conclude by saying that it's a real and true thing that it can be hard to work with derived signals built from (non-local) resources, etc. I think it's mostly because it is a genuinely really complicated task — starting some async task on the server, resolving it on the client, using the data from that and passing it through context and accessing it anywhere in your component... this is super cool and also tricky! 

There are cases in which swiching to a local resource can be appropriate: it's a performance deopt in a real sense (it doesn't start loading until WASM has loaded, etc.) but in some cases that's not make or break, and it can make the whole Suspense/hydration system less brittle.

**Comment by gbj** [comment]
Just a short update to say: Something very close to this is planned for 0.6.

**Comment by haslersn** [comment]
> ## When you _can_ read a resource without Suspense
> [...]
>     3. using `create_local_resource`

I noticed that calling `loading()` on a local resource on the server (with SSR) always returns `false`. I didn't know about Transition/Suspense, yet, so I used `if resource.loading() { ... }` instead and was surprised that the `else` branch is initially rendered. There also were some hydration bugs related to this.

This leads to two suggestions:

- Make `loading()` on local resources always return `true` on the server. (This matches that `get()` returns None.)
- Print a warning when a local resource is accessed (e.g., `loading()` or `get()` is called) on the server.

**Comment by rakshith-ravi** [comment]
@haslersn - If you take a look at here: https://docs.rs/leptos/latest/leptos/fn.create_local_resource.html, local resources are always run on the client side, so I suppose _technically_ it is not loading, but yes, I do see your point about having it return true on the server

**Comment by gbj** [comment]
In 0.7/current main branch, top-level Suspend without a Suspense/Transition will create its own Suspense wrapper. The Render implementation for resources creates a Suspend. I would still recommend using an actual Suspense in most cases, rather than a bunch of individual resources that are streamed in separately.
